src/storage/ring_buffer.rs

```diff
@@ -30,6 +30,24 @@ impl<'a, T: 'a> RingBuffer<'a, T> {
         }
     }
 
+    /// Create a ring buffer with the given storage.
+    ///
+    /// During creation, every element in `storage` is set to Default::default().
+    pub fn with_default<S>(storage: S) -> RingBuffer<'a, T>
+        where S: Into<ManagedSlice<'a, T>>, T: Default,
```

Comment:
I'm uncomfortable about this ambiguity, having two separate paths for resetting and creating a default member. How about making `Resettable` imply `Default` instead?

Reply:
I don't think we can do it, otherwise, I would have gone with `Default` when I refactored `RingBuffer` out. Currently, we're using `RingBuffer` to store `PacketBuffer` for UDP and Raw sockets, and these contain `Managed<'a, [u8]>`, meaning we can't just `Default::default()` them, so we use `Resettable` to reset the other fields. Yet for 90% of the potential use cases `Default::default()` would be enough and there's no need to require `Resettable`.

Now I think of it, maybe we should just get rid of `Resettable` and initialization of elements in the constructors altogether? I can't remember why I even need them, the container itself doesn't care about the values, and we already initialize them in the constructors before inserting.

Reply:
> Now I think of it, maybe we should just get rid of Resettable and initialization of elements in the constructors altogether?

I think you'll get infoleaks when reusing sockets then.

Reply:
We only reset elements in the `RingBuffer`'s constructor tho, `try_enqueue` explicitly takes a function to reset the new element, and `enqueue` doesn't do it at all.

Reply:
Exactly. Suppose application A put key material all over the socket. Then, the socket, without being reset, is passed to application B. This application can now call `.enqueue()` and read from the buffer what it should have never seen. Ideally we'd have something like `&write [u8]` but alas.

Reply:
This is an interesting point I didn't think about, but I think we should just make `RingBuffer`'s methods pub(crate) since a user doesn't need to use them directly. It's impossible to get access to the old data via sockets' methods because they all limited by the length of the data in the buffer. Besides, I think memsetting buffers all the time would be too costly.

Reply:
Of course it's possible, just use `tcp_socket.send()`, which directly calls `enqueue()`...

Reply:
`TcpSocket` doesn't use `RingBuffer` right now, but you're right, you can get access to the old data via `UdpSocket::send()`. So the current problem is twofold - we don't call `reset` on element reuse and we don't memset the buffer in `reset`. I guess I'll do it then. This means the `Resettable` bound is now container-wide and there's no point in `with_default` constructor.

Now I need to find a way to memset a buffer in such a way that the compiler won't optimize it out :)

Reply:
It's only legal for memset to be removed right before deallocation, and we don't have that, so it's not a concern.